### ISSUE TITLE
np.random.multinomial throws ValueError: binomial(): p outside of [0, 1]

### ISSUE THREAD

**Issue by al5250** [numba/numba#3426]
## Reporting a bug

- [ ] I am using the latest released version of Numba (most recent is visible in
 the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).
- [ ] I have included below a minimal working reproducer (if you are unsure how
 to write one see http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports).

Suppose I have a vector of probabilities that add up to 1.  Sometimes, I have found that using `np.random.multinomial` throws an error.  For example,
```
A = np.array([0.     , 0.91954, 0.01867, 0.     , 0.     , 0.01745, 0.04434,
              0.     , 0.     , 0.     , 0.     , 0.     ])
@jit(nopython=True)
def samp_multi(x):
    return np.random.multinomial(1, x)

for _ in range(100):
    samp_multi(A)
```
leads to the following error:

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-794-f6a667d67d4f> in <module>()
      1 for _ in range(100):
----> 2     samp_multi(A)

//anaconda/envs/py35/lib/python3.5/site-packages/numba/targets/randomimpl.py in binomial_impl()
    812             raise ValueError("binomial(): n <= 0")
    813         if not (0.0 <= p <= 1.0):
--> 814             raise ValueError("binomial(): p outside of [0, 1]")
    815         if p == 0.0:
    816             return 0

ValueError: binomial(): p outside of [0, 1]
```

**Comment by rjenc29** [comment]
Just a thought: the fraction **p_j / p_sum** may be affected by float precision behaviour (numba/targets/randomimpl.py:1444):

```python
n_j = fl[i + j] = np.random.binomial(n_experiments, p_j / p_sum)
```
If so, perhaps capping the ratio at 1 would make sense / be justified:

```python
n_j = fl[i + j] = np.random.binomial(n_experiments, min(p_j / p_sum, 1))
```

**Comment by stuartarchibald** [comment]
Thanks for the report. I think Numba needs to follow NumPy better here and compute the more stable Kahan sum when assessing this condition (and also check out the numerical stability of the call to `np.random.binomial` as @rjenc29 suggests).

https://github.com/numpy/numpy/blob/db5750f6cdc2715f1c65be31f985e2cd2699d2e0/numpy/random/mtrand/mtrand.pyx#L4627-L4628

Marking this as a bug.

**Comment by lrsppp** [comment]
I could find a work around by applying `np.round(p, decimal)` on the probability vector first. I used `decimal = 12` and it worked fine. Increasing `decimal` further lead to the same error/bug again.

**Comment by juliusbierk** [comment]
Bump! :)

Just can into the same issue:

```
import numba as nb
import numpy as np


def multinomial(ps):
    return np.random.multinomial(2, ps)


ps = np.array([0.4166666666666667, 0.5833333333333333, 0.0])
print(multinomial(ps))  # works
print(nb.njit(multinomial)(ps))  # works

ps = np.array([0.4166666666666667, 0.5833333333333334, 0.0])
print(multinomial(ps))  # works
print(nb.njit(multinomial)(ps))  # fails

```

**Comment by juliusbierk** [comment]
> Just a thought: the fraction **p_j / p_sum** may be affected by float precision behaviour (numba/targets/randomimpl.py:1444):
> 
> ```python
> n_j = fl[i + j] = np.random.binomial(n_experiments, p_j / p_sum)
> ```
> 
> If so, perhaps capping the ratio at 1 would make sense / be justified:
> 
> ```python
> n_j = fl[i + j] = np.random.binomial(n_experiments, min(p_j / p_sum, 1))
> ```

IMO this seems right.
The only issue could be if the user supplies probabilities whose sum exceed one.
They should then be warned. Perhaps and error?

**Comment by juliusbierk** [comment]
I tried to submit a suggestive PR.
Should the behaviour also be tested?

**Comment by esc** [comment]
> I tried to submit a suggestive PR. Should the behaviour also be tested?

Yes, we do like tests, especially for this kind of stuff. The PR is #7551 ?

**Comment by juliusbierk** [comment]
> > I tried to submit a suggestive PR. Should the behaviour also be tested?
> 
> Yes, we do like tests, especially for this kind of stuff. The PR is #7551 ?

Yes that is the PR. I just added a test.

**Comment by georgewambold** [comment]
Bumping this -- Just ran into the same `ValueError: binomial(): p outside of [0, 1]` issue. The code runs without issue in python + object mode, but `nopython=True` causes an error to be raised.

**Comment by njwfish** [comment]
Bumping again here in 2025 -- it would be great if we could merge even the minimal `min(p_j / p_sum, 1)` fix for this. When I edit that into source it resolves my issues, as discussed [here](https://github.com/numba/numba/pull/7551).

I appreciate the interest in making numba and numpy compatible and numpy's implementation is clearly a better more stable approach, but I really think while work on `np.random.Generator` continues it'd be great to fold in something like [PR7751](https://github.com/numba/numba/pull/7551).

**Comment by juliusbierk** [comment]
@njwfish I agree. Still happy to try to get that PR merged if someone from numba agrees :)

**Comment by simozacca** [comment]
Would it be possible to please fix this? Thank you!

**Comment by esc** [comment]
> Would it be possible to please fix this? Thank you!

Thank you for commenting on this issue. Unfortunately the project is very resource constrained so there is currently no ETA on getting this fixed as there are no maintainers available for reviewing PRs. This year, the maintainer time has been largely spent on porting the CI system to GitHub actions (and updating the release process) as well as implementing support for updated core dependencies: NumPy (`2.*`), CPython (`3.14`) and LLVM (`20.*`). Though I am hopeful that next year (2026) more maintainer time can be allocated to servicing contribution PRs and fixing bugs such as this one. 

The best way to help is to become a maintainer who is in a position to review PRs, because that is currently the bottleneck. I have created some preliminary instructions on how to do that here:

https://github.com/numba/numba/wiki/Numba-Maintainance
